src/Microsoft.OData.Client/TypeBuilderUtil.cs

```diff
@@ -0,0 +1,107 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="TypeBuilderUtil.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Client
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+    using System.Reflection.Emit;
+
+
+    internal static class TypeBuilderUtil
+    {
+        /// <summary>
+        /// Dynamically builds a type from the key/value pairs in the dictionary 
+        /// where key is the property name and value is the property type.
+        /// </summary>
+        /// <param name="typeName">The name of the created type.</param>
+        /// <param name="properties">Dictionary containing properties and their respective types.</param>
+        /// <returns></returns>
+        public static TypeInfo CreateTypeInfo(string typeName, IDictionary<string, Type> properties)
+        {
+            TypeBuilder typeBuilder = GetTypeBuilder(typeName);
+
+            foreach (KeyValuePair<string, Type> kvPair in properties)
+            {
+                CreateProperty(typeBuilder, kvPair.Key, kvPair.Value);
+            }
+
+            return typeBuilder.CreateTypeInfo();
+        }
+
+        /// <summary>
+        /// Returns type builder.
+        /// </summary>
+        /// <param name="typeName">The typename.</param>
+        /// <returns></returns>
+        private static TypeBuilder GetTypeBuilder(string typeName)
+        {
+            string assembyName = Guid.NewGuid().ToString().Replace("-", "").Substring(0, 16);
+
+            // Parameterize AssemblyBuilderAccess value if that ever becomes necessary
+            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(
+                new AssemblyName(assembyName), AssemblyBuilderAccess.RunAndCollect);

```

Comment:
I think it would be interesting at some point to understand the lifetime of the generated assembly and when it's actually collected, and its impact on resources.

If an aggregation query that returns the same properties is called multiple times, could we consider caching the dynamically generated type instead of building it afresh on every call of the query?

Reply:
@habbes About lifetime of the generated assembly, this [document](https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/collectible-assemblies) could explain it better.
About caching the generated type, we'd need to cache the assembly it belongs to, rather than the type. At the moment, I don't know if that's doable, or easy or hard to accomplish, or what effect if any that would have on the assembly's "collectability". I could spend time exploring that but it'd make more sense to first conduct some tests to estimate the cost we incur from generating the collectible assembly.